README.rst
collective.angularstarter
=========================

Get started with Plone + AngularJS (Yeoman powered) without any of the normal headaches associated with a manual setup.

Useful when you want to develop a single page web applications powered by Angular using Plone as backend.

How it works
------------

Before installing collective.angularstarter you need to:

* # install nodejs (suggested way NVM)
* $ npm install -g grunt-cli bower
* $ cd collective.angularstarter/collective/angularstarter/browser/angular
* $ bower install
* $ npm install
* $ grunt build

After that install the collective.angularstarter plugin and choose the more appropriate install profile:

* :default -> for production (htmlmin, image minification, concat+uglify/cssmin, uncss, cdnify)
* :devel -> suitable for development

Now you can visit your AngularJS single page web application:

* http://localhost:8080/Plone/@@angularstarter


.. figure:: https://raw.github.com/collective/collective.angularstarter/master/docs/screenshots/angularplone.png
    :figwidth: image

     Plone + Angular kickstarter. Bootstrap your AngularJS based single webapp applications with collective.angularstarter

You can extend this project and then clone it with::

    $ VENV/bin/collective_angularstarter_clone your.awesomeplugin

TODO
----

* test production vs devel layers
* makefile for init project

Links
-----

* for further details see http://davidemoro.blogspot.com/2014/09/plone-angularjs-yeoman-starter.html

Authors
-------

* Davide Moro <davide.moro@gmail.com> | @davidemoro | http://davidemoro.blogspot.it/



docs/HISTORY.txt
Changelog
=========

0.1dev (unreleased)
-------------------

- Initial release
  [@davidemoro]


collective/angularstarter/tests.py
import unittest

#from zope.testing import doctestunit
#from zope.component import testing
from Testing import ZopeTestCase as ztc

from Products.Five import fiveconfigure
from Products.PloneTestCase import PloneTestCase as ptc
from Products.PloneTestCase.layer import PloneSite
ptc.setupPloneSite()

import collective.angularstarter


class TestCase(ptc.PloneTestCase):

    class layer(PloneSite):

        @classmethod
        def setUp(cls):
            fiveconfigure.debug_mode = True
            ztc.installPackage(collective.angularstarter)
            fiveconfigure.debug_mode = False

        @classmethod
        def tearDown(cls):
            pass


def test_suite():
    return unittest.TestSuite([

        # Unit tests
        #doctestunit.DocFileSuite(
        #    'README.txt', package='collective.angularstarter',
        #    setUp=testing.setUp, tearDown=testing.tearDown),

        #doctestunit.DocTestSuite(
        #    module='collective.angularstarter.mymodule',
        #    setUp=testing.setUp, tearDown=testing.tearDown),


        # Integration tests that use PloneTestCase
        #ztc.ZopeDocFileSuite(
        #    'README.txt', package='collective.angularstarter',
        #    test_class=TestCase),

        #ztc.FunctionalDocFileSuite(
        #    'browser.txt', package='collective.angularstarter',
        #    test_class=TestCase),

        ])

if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


collective/angularstarter/browser/angular/test/karma.conf.js
// Karma configuration
// http://karma-runner.github.io/0.12/config/configuration-file.html
// Generated on 2014-09-09 using
// generator-karma 0.8.3

module.exports = function(config) {
  'use strict';

  config.set({
    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,

    // base path, that will be used to resolve files and exclude
    basePath: '../',

    // testing framework to use (jasmine/mocha/qunit/...)
    frameworks: ['jasmine'],

    // list of files / patterns to load in the browser
    files: [
      '++resource++collective.angularstarter.bower_components/angular/angular.js',
      '++resource++collective.angularstarter.bower_components/angular-mocks/angular-mocks.js',
      '++resource++collective.angularstarter.bower_components/angular-animate/angular-animate.js',
      '++resource++collective.angularstarter.bower_components/angular-cookies/angular-cookies.js',
      '++resource++collective.angularstarter.bower_components/angular-resource/angular-resource.js',
      '++resource++collective.angularstarter.bower_components/angular-route/angular-route.js',
      '++resource++collective.angularstarter.bower_components/angular-sanitize/angular-sanitize.js',
      '++resource++collective.angularstarter.bower_components/angular-touch/angular-touch.js',
      '++resource++collective.angularstarter.bower_components/angular-translate/angular-translate.js',
      'app/scripts/**/*.js',
      'test/mock/**/*.js',
      'test/spec/**/*.js'
    ],

    // list of files / patterns to exclude
    exclude: [],

    // web server port
    port: 8080,

    // Start these browsers, currently available:
    // - Chrome
    // - ChromeCanary
    // - Firefox
    // - Opera
    // - Safari (only Mac)
    // - PhantomJS
    // - IE (only Windows)
    browsers: [
      'PhantomJS'
    ],

    // Which plugins to enable
    plugins: [
      'karma-phantomjs-launcher',
      'karma-chrome-launcher',
      'karma-jasmine'
    ],

    // Continuous Integration mode
    // if true, it capture browsers, run tests and exit
    singleRun: false,

    colors: true,

    // level of logging
    // possible values: LOG_DISABLE || LOG_ERROR || LOG_WARN || LOG_INFO || LOG_DEBUG
    logLevel: config.LOG_INFO,

    // Uncomment the following lines if you are using grunt's server to run the tests
    // proxies: {
    //   '/': 'http://localhost:9000/'
    // },
    // URL root prevent conflicts with the site root
    // urlRoot: '_karma_'
  });
};


collective/angularstarter/browser/angular/test/spec/controllers/main.js
'use strict';

describe('Controller: MainCtrl initialization', function () {

  // load the controller's module
  beforeEach(module('angularApp'));

  var MainCtrl,
    scope;

  // Initialize the controller and a mock scope
  beforeEach(inject(function ($controller, $rootScope, $q) {
    scope = $rootScope.$new();
    MainCtrl = $controller('MainCtrl', {
      $scope: scope,
    });
  }));

  it('message should be Plone!', function () {
    expect(scope.message).toBe('Plone!');
  });
});



collective/angularstarter/scripts.py
import sys
import os
import string
import shutil
import optparse
import textwrap
import pkg_resources

def collective_angularstarter_clone():
    description = """\
    Print the deployment settings for a Pyramid application.  Example:
    'show_settings deployment.ini'
    """
    usage = "usage: collective_angularstarter_clone new_package_name"
    parser = optparse.OptionParser(
        usage=usage,
        description=textwrap.dedent(description)
        )

    options, args = parser.parse_args(sys.argv[1:])
    if not len(args) == 1:
        print('You must provide at least one argument')
        return 2

    print('Start translating package...')
    _translate(args[0])
    print('Done!')


def _translate(new_package_name, original_package_name='collective.angularstarter'):
    """ Translate package """

    def package_split(package_name, separator='.'):
        """ With 'pippo.pluto.paperino' you'll get:
            ['pippo.pluto.paperino','pippo.pluto', 'pippo', 'pluto', 'paperino']
        """
        items = package_name.split('.')
        len_items = len(items)
        results = [package_name, items[-1].capitalize()]
    
        for count in range(1, len_items-1):
            el = items[:-count]
            if el:
                results.append('.'.join(el))   
        results.extend(items)
        return results
    
    # copy tree
    original_location = pkg_resources.get_distribution(original_package_name).location
    shutil.copytree(original_location,
                    new_package_name,
                    ignore=shutil.ignore_patterns('*.pyc', '.svn', '*.tmp', '*.egg-info', '*.egg' 
                                                  '.git', 'node_modules', '*.swp', '*~',))
    
    original_slices = package_split(original_package_name)
    new_slices = package_split(new_package_name)
    trans_map = zip(original_slices, new_slices)
    for step in os.walk(top=new_package_name, topdown=False):
        item_path = step[0]
        item_dir_paths = step[1]
        item_file_paths = step[2]
    
        for item_file_path in item_file_paths:
            # read the file and substitute patterns
            file_contents = ''
            # TODO: avoid rewriting of '*.png', '*.gif', '*.svg', '*.ico', '*.jpg'
            # TODO: avoid duplicated open files
            with open(os.path.join(item_path, item_file_path), 'r') as file_to_filter:
                file_contents = file_to_filter.read()

            # translate file
            for trans in trans_map:
                old = trans[0]
                new = trans[1]
                file_contents = string.replace(file_contents, old, new) 
                file_contents = string.replace(file_contents, old.capitalize(), new.capitalize()) 

            # write translated version of file
            with open(os.path.join(item_path, item_file_path), 'w') as file_to_filter:
                file_to_filter.write(file_contents)
    
            # rename files
            for trans in trans_map: 
                old = trans[0]
                new = trans[1]
                if old in item_file_path:
                    old_path = item_file_path
                    new_path = string.replace(old_path, old, new)
                    os.rename(os.path.join(item_path, old_path), os.path.join(item_path, new_path))
                    break
                    
        for item_dir_path in item_dir_paths:
            for trans in trans_map: 
                old = trans[0]
                new = trans[1]
                if old in item_dir_path:
                    old_path = item_dir_path
                    new_path = string.replace(old_path, old, new)
                    os.rename(os.path.join(item_path, old_path), os.path.join(item_path, new_path))
                    break
    


collective/angularstarter/browser/angular/app/++resource++collective.angularstarter.scripts/app.js
'use strict';

/**
 * @ngdoc overview
 * @name angularApp
 * @description
 * # angularApp
 *
 * Main module of the application.
 */
angular
  .module('angularApp', [
    'pascalprecht.translate',
    'ngRoute'
  ])
  .config(['$translateProvider', function ($translateProvider) {
    var defaultLanguage = 'en';

    $translateProvider.translations('en', {
      HELLO: 'Hello'
    });
    $translateProvider.translations('it', {
      HELLO: 'Ciao',
    });
    $translateProvider.preferredLanguage(defaultLanguage);
    $translateProvider.determinePreferredLanguage(function () {
      // is navigator language ok or should we consider also the user's lang cookie
      // since we are dealing with plone?
      var navigatorLanguage = navigator.language;

      return navigatorLanguage ? navigatorLanguage : defaultLanguage;
    });
  }])
  .config(['$routeProvider', function ($routeProvider) {
    $routeProvider
      .when('/main', {
        templateUrl: '++resource++collective.angularstarter.views/main.html',
        controller: 'MainCtrl'
      })
      .otherwise({
        redirectTo: '/main'
      });
  }]);


collective/angularstarter/browser/angular/Gruntfile.js
// Generated on 2014-09-09 using generator-angular 0.9.8
'use strict';

// # Globbing
// for performance reasons we're only matching one level down:
// 'test/spec/{,*/}*.js'
// use this if you want to recursively match all subfolders:
// 'test/spec/**/*.js'

module.exports = function (grunt) {

  // Load grunt tasks automatically
  require('load-grunt-tasks')(grunt);

  // Time how long tasks take. Can help when optimizing build times
  require('time-grunt')(grunt);

  // Configurable paths for the application
  var appConfig = {
    app: require('./bower.json').appPath || 'app',
    tmp: '.tmp',
    dist: 'dist'
  };

  // Define the configuration for all the tasks
  grunt.initConfig({

    // Project settings
    yeoman: appConfig,

    // Watches files for changes and runs tasks based on the changed files
    watch: {
      bower: {
        files: ['bower.json'],
        tasks: ['wiredep']
      },
      js: {
        files: ['<%= yeoman.app %>/++resource++collective.angularstarter.scripts/{,*/}*.js'],
        tasks: ['newer:jshint:all'],
        options: {
          livereload: '<%= connect.options.livereload %>'
        }
      },
      jsTest: {
        files: ['test/spec/{,*/}*.js'],
        tasks: ['newer:jshint:test', 'karma']
      },
      styles: {
        files: ['<%= yeoman.app %>/++resource++collective.angularstarter.styles/{,*/}*.css'],
        tasks: ['newer:copy:styles', 'autoprefixer']
      },
      gruntfile: {
        files: ['Gruntfile.js']
      },
      livereload: {
        options: {
          livereload: '<%= connect.options.livereload %>'
        },
        files: [
          '<%= yeoman.app %>/{,*/}*.html',
          '.tmp/++resource++collective.angularstarter.styles/{,*/}*.css',
          '<%= yeoman.app %>/++resource++collective.angularstarter.images/{,*/}*.{png,jpg,jpeg,gif,webp,svg}'
        ]
      }
    },

    // The actual grunt server settings
    connect: {
      options: {
        port: 9000,
        // Change this to '0.0.0.0' to access the server from outside.
        hostname: 'localhost',
        livereload: 35729
      },
      livereload: {
        options: {
          open: true,
          middleware: function (connect) {
            return [
              connect.static('.tmp'),
              connect().use(
                '/++resource++collective.angularstarter.bower_components',
                connect.static('./++resource++collective.angularstarter.bower_components')
              ),
              connect.static(appConfig.app)
            ];
          }
        }
      },
      test: {
        options: {
          port: 9001,
          middleware: function (connect) {
            return [
              connect.static('.tmp'),
              connect.static('test'),
              connect().use(
                '/++resource++collective.angularstarter.bower_components',
                connect.static('./++resource++collective.angularstarter.bower_components')
              ),
              connect.static(appConfig.app)
            ];
          }
        }
      },
      dist: {
        options: {
          open: true,
          base: '<%= yeoman.dist %>'
        }
      }
    },

    // Make sure code styles are up to par and there are no obvious mistakes
    jshint: {
      options: {
        jshintrc: '.jshintrc',
        reporter: require('jshint-stylish')
      },
      all: {
        src: [
          'Gruntfile.js',
          '<%= yeoman.app %>/++resource++collective.angularstarter.scripts/{,*/}*.js'
        ]
      },
      test: {
        options: {
          jshintrc: 'test/.jshintrc'
        },
        src: ['test/spec/{,*/}*.js']
      }
    },

    // Empties folders to start fresh
    clean: {
      dist: {
        files: [{
          dot: true,
          src: [
            '.tmp',
            '<%= yeoman.dist %>/{,*/}*',
            '!<%= yeoman.dist %>/.git*'
          ]
        }]
      },
      server: '.tmp'
    },

    // Add vendor prefixed styles
    autoprefixer: {
      options: {
        browsers: ['last 1 version']
      },
      dist: {
        files: [{
          expand: true,
          cwd: '.tmp/++resource++collective.angularstarter.styles/',
          src: '{,*/}*.css',
          dest: '.tmp/++resource++collective.angularstarter.styles/'
        }]
      }
    },

    // Automatically inject Bower components into the app
    wiredep: {
      app: {
        src: ['<%= yeoman.app %>/index.html'],
        exclude: ['./\\+\\+resource\\+\\+collective.angularstarter.bower_components/bootstrap/dist/js/bootstrap.js', './\\+\\+resource\\+\\+collective.angularstarter.bower_components/jquery/dist/jquery.js', './\\+\\+resource\\+\\+collective.angularstarter.bower_components/angular-resource/angular-resource.js', '.\\+\\+resource\\+\\+collective.angularstarter.bower_components/angular-cookies/angular-cookies.js', './\\+\\+resource\\+\\+collective.angularstarter.bower_components/angular-sanitize/angular-sanitize.js', './\\+\\+resource\\+\\+collective.angularstarter.bower_components/angular-animate/angular-animate.js', './\\+\\+resource\\+\\+collective.angularstarter.bower_components/angular-touch/angular-touch.js'],
//        exclude: ['./++resource++collective.angularstarter.bower_components/bootstrap/dist/js/bootstrap.js', './++resource++collective.angularstarter.bower_components/jquery/dist/jquery.js', './++resource++collective.angularstarter.bower_components/angular-resource/angular-resource.js', '.++resource++collective.angularstarter.bower_components/angular-cookies/angular-cookies.js', './++resource++collective.angularstarter.bower_components/angular-sanitize/angular-sanitize.js', './++resource++collective.angularstarter.bower_components/angular-animate/angular-animate.js', './++resource++collective.angularstarter.bower_components/angular-touch/angular-touch.js'],
        ignorePath:  /\.\.\//
      }
    },

    // Renames files for browser caching purposes
    filerev: {
      dist: {
        src: [
          '<%= yeoman.dist %>/++resource++collective.angularstarter.scripts/{,*/}*.js',
          '<%= yeoman.dist %>/++resource++collective.angularstarter.styles/{,*/}*.css',
          '<%= yeoman.dist %>/++resource++collective.angularstarter.images/{,*/}*.{png,jpg,jpeg,gif,webp,svg}',
          '<%= yeoman.dist %>/++resource++collective.angularstarter.fonts/*'
        ]
      }
    },

    // Reads HTML for usemin blocks to enable smart builds that automatically
    // concat, minify and revision files. Creates configurations in memory so
    // additional tasks can operate on them
    useminPrepare: {
      html: '<%= yeoman.app %>/index.html',
      options: {
        dest: '<%= yeoman.dist %>',
        flow: {
          html: {
            steps: {
              js: ['concat', 'uglifyjs'],
              css: ['concat', 'cssmin']
            },
            post: {}
          }
        }
      }
    },

    // Performs rewrites based on filerev and the useminPrepare configuration
    usemin: {
      html: ['<%= yeoman.dist %>/{,*/}*.html'],
      css: ['<%= yeoman.dist %>/++resource++collective.angularstarter.styles/{,*/}*.css'],
      options: {
        assetsDirs: ['<%= yeoman.dist %>','<%= yeoman.dist %>/++resource++collective.angularstarter.images']
      }
    },

    // The following *-min tasks will produce minified files in the dist folder
    // By default, your `index.html`'s <!-- Usemin block --> will take care of
    // minification. These next options are pre-configured if you do not wish
    // to use the Usemin blocks.
    // cssmin: {
    //   dist: {
    //     files: {
    //       '<%= yeoman.dist %>/++resource++collective.angularstarter.styles/main.css': [
    //         '.tmp/++resource++collective.angularstarter.styles/{,*/}*.css'
    //       ]
    //     }
    //   }
    // },
    // uglify: {
    //   dist: {
    //     files: {
    //       '<%= yeoman.dist %>/++resource++collective.angularstarter.scripts/scripts.js': [
    //         '<%= yeoman.dist %>/++resource++collective.angularstarter.scripts/scripts.js'
    //       ]
    //     }
    //   }
    // },
    // concat: {
    //   dist: {}
    // },

    uncss: {
      dist: {
        options: {
          ignore: [/li\.active \> a/],
          csspath: '../.tmp',
          stylesheets: ['concat/++resource++collective.angularstarter.styles/main.css']
        },
        files: {
          '.tmp/concat/++resource++collective.angularstarter.styles/main.css': ['<%= yeoman.dist %>/index.html', '<%= yeoman.dist %>/++resource++collective.angularstarter.views/main.html', '<%= yeoman.dist %>/++resource++collective.angularstarter.views/searchbycode.html']
        }
      }
    },

    imagemin: {
      dist: {
        files: [{
          expand: true,
          cwd: '<%= yeoman.app %>/images',
          src: '{,*/}*.{png,jpg,jpeg,gif}',
          dest: '<%= yeoman.dist %>/++resource++collective.angularstarter.images'
        }]
      }
    },

    svgmin: {
      dist: {
        files: [{
          expand: true,
          cwd: '<%= yeoman.app %>/++resource++collective.angularstarter.images',
          src: '{,*/}*.svg',
          dest: '<%= yeoman.dist %>/++resource++collective.angularstarter.images'
        }]
      }
    },

    htmlmin: {
      dist: {
        options: {
          collapseBooleanAttributes: true,
          collapseWhitespace: true,
          removeComments: true,
          removeCommentsFromCDATA: true,
          removeEmptyAttributes: true,
          removeRedundantAttributes: true,
          useShortDoctype: true
        },
        files: [{
          expand: true,
          cwd: '<%= yeoman.dist %>',
          src: ['*.html', '++resource++collective.angularstarter.views/{,*/}*.html'],
          dest: '<%= yeoman.dist %>'
        }]
      }
    },

    // ng-annotate tries to make the code safe for minification automatically
    // by using the Angular long form for dependency injection.
    ngAnnotate: {
      dist: {
        files: [{
          expand: true,
          cwd: '.tmp/concat/scripts',
          src: ['*.js', '!oldieshim.js'],
          dest: '.tmp/concat/scripts'
        }]
      }
    },

    // Replace Google CDN references
    cdnify: {
      dist: {
        html: ['<%= yeoman.dist %>/*.html']
      }
    },

    sed: {
      html: {
      path: '<%= yeoman.dist %>/index.html',
      pattern: '\\+\\+resource\\+\\+collective.angularstarter.\\/\\/',
      replacement: '//'}
    },

    // Copies remaining files to places other tasks can use
    copy: {
      dist: {
        files: [{
          expand: true,
          dot: true,
          cwd: '<%= yeoman.app %>',
          dest: '<%= yeoman.dist %>',
          src: [
            '*.{ico,png,txt}',
            '.htaccess',
            '*.html',
            '++resource++collective.angularstarter.views/{,*/}*.html',
            '++resource++collective.angularstarter.images/{,*/}*.{webp}',
            '++resource++collective.angularstarter.fonts/*'
          ]
        }, {
          expand: true,
          cwd: '.tmp/++resource++collective.angularstarter.images',
          dest: '<%= yeoman.dist %>/++resource++collective.angularstarter.images',
          src: ['generated/*']
        }]
      },
      styles: {
        expand: true,
        cwd: '<%= yeoman.app %>/styles',
        dest: '.tmp/++resource++collective.angularstarter.styles/',
        src: '{,*/}*.css'
      }
    },

    // Run some tasks in parallel to speed up the build process
    concurrent: {
      server: [
        'copy:styles'
      ],
      test: [
        'copy:styles'
      ],
      dist: [
        'copy:styles',
        'imagemin',
        'svgmin'
      ]
    },

    // Test settings
    karma: {
      unit: {
        configFile: 'test/karma.conf.js',
        singleRun: true
      }
    }
  });


  grunt.registerTask('serve', 'Compile then start a connect web server', function (target) {
    if (target === 'dist') {
      return grunt.task.run(['build', 'connect:dist:keepalive']);
    }

    grunt.task.run([
      'clean:server',
      'wiredep',
      'concurrent:server',
      'autoprefixer',
      'connect:livereload',
      'watch'
    ]);
  });

  grunt.registerTask('server', 'DEPRECATED TASK. Use the "serve" task instead', function (target) {
    grunt.log.warn('The `server` task has been deprecated. Use `grunt serve` to start a server.');
    grunt.task.run(['serve:' + target]);
  });

  grunt.registerTask('test', [
    'clean:server',
    'concurrent:test',
    'autoprefixer',
    'connect:test',
    'karma'
  ]);

  grunt.registerTask('build', [
    'clean:dist',
    'wiredep',
    'useminPrepare',
    'concurrent:dist',
    'autoprefixer',
    'concat',
    'ngAnnotate',
    'copy:dist',
    'uncss',
    'cdnify',
    'sed',   // See https://github.com/btford/grunt-google-cdn/issues/36
    'cssmin',
    'uglify',
    'filerev',
    'usemin',
    'htmlmin'
  ]);

  grunt.registerTask('default', [
    'newer:jshint',
    'test',
    'build'
  ]);
};
